[lldb] [debugserver] fix qLaunchSuccess error, add QErrorStringInPacketSupported

jasonmolenda opened the pull request:
Pavel added an extension to lldb's gdb remote serial protocol that allows the debug stub to append an error message (ascii hex encoded) after an error response packet Exx.  This was added in 2017 in https://reviews.llvm.org/D34945 .  lldb sends the QErrorStringInPacketSupported packet and then the remote stub may add these error strings.

debugserver has two bugs in its use of extended error messages:  the vAttach family would send the extended error string without checking if the mode had been enabled.  And qLaunchSuccess would not properly format its error response packet (missing the hex digits, did not asciihex encode the string).

There is also a bug in the HandlePacket_D (detach) packet where the error packets did not include hex digits, but this one does not append an error string.

I'm adding a new RNBRemote::SendErrorPacket() and routing all error packet returns though this one method.  It takes an optional second string which is the longer error message; it now handles appending it to the Exx response or not, depending on the QErrorStringInPacketSupported state.  I updated all packets to send their errors via this method.

jasonmolenda committed:
[lldb] [debugserver] fix qLaunchSuccess error, add QErrorStringInPacketSupported

Pavel added an extension to lldb's gdb remote serial protocol that
allows the debug stub to append an error message (ascii hex encoded)
after an error response packet Exx.  This was added in 2017 in
https://reviews.llvm.org/D34945 .  lldb sends the
QErrorStringInPacketSupported packet and then the remote stub may
add these error strings.

debugserver has added these strings to the vAttach family of packets
to explain why an attach failed, without waiting to see the
QErrorStringInPacketSupported packet to enable the mode.

debugserver also has a bad implementation of this for the qLaunchSuccess
packet, where "E" is sent followed by the literal characters of the
error, instead of Exx;<asciihex string>, which lldb can have problems
parsing.

This patch moves three utility functions earlier in RNBRemote.cpp,
it accepts the QErrorStringInPacketSupported packet and only appends
the expanded error messages when that has been sent (lldb always sends
it at the beginning of a connection), fixes the error string returned
by qLaunchSuccess and changes the vAttach error returns to only add
the error string when the mode is enabled.

bulbazord commented on lldb/tools/debugserver/source/RNBRemote.cpp:
```diff
@@ -3944,15 +3955,22 @@ rnb_err_t RNBRemote::HandlePacket_v(const char *p) {
           DNBLogError("Tried to attach to pid that doesn't exist");
-          std::string return_message = "E96;";
-          return_message += cstring_to_asciihex_string("no such process.");
+          std::string return_message = "E96";
```
I see this pattern in lots of places, is it worth abstracting out? Something like:

```
std::string CreateAttachError(const char *additional_context = nullptr) {
    std::string message = "E96";
    if (additional_context)
        message += ";" + cstring_to_asciihex_string(additional_context);
    return message;
}
```

What do you think?

clayborg replied:
Then this line becomes:
```
std::string return_message = CreateEXXPacket("E96", "no such process");
```

jasonmolenda replied:
I'm thinking of having a SendErrorPacket() instead of SendPacket() with an error string optional argument, and having all error packets route through that, and having all SendPacket("Exx")'s go through that instead.  @bulbazord @clayborg what do you think?  I'm having trouble deciding if I want to have a method that constructs the error string and the caller passes it to SendPacket or if I want to have a method to do both.

clayborg replied:
> I'm thinking of having a SendErrorPacket() instead of SendPacket() with an error string optional argument, and having all error packets route through that, and having all SendPacket("Exx")'s go through that instead. @bulbazord @clayborg what do you think? I'm having trouble deciding if I want to have a method that constructs the error string and the caller passes it to SendPacket or if I want to have a method to do both.

I like that idea!

bulbazord replied:
Sounds great to me 😄

bulbazord reviewed (approved):
LGTM, seems like a good fix. One question/suggestion.

clayborg commented on lldb/tools/debugserver/source/RNBRemote.cpp:
```diff
@@ -3969,30 +3987,43 @@ rnb_err_t RNBRemote::HandlePacket_v(const char *p) {
+          std::string return_message = "E96";
+          if (m_enable_error_strings) {
+            return_message += ";";
+            std::string msg = "tried to attach to process as user '" +
+                              my_username +
+                              "' and process is running "
+                              "as user '" +
+                              process_username + "'";
```
not ended with a '.' here, we pick how we want to do it and do it consistently

clayborg commented on lldb/tools/debugserver/source/RNBRemote.cpp:
```diff
@@ -3969,30 +3987,43 @@ rnb_err_t RNBRemote::HandlePacket_v(const char *p) {
+          if (m_enable_error_strings) {
+            return_message += ";";
+            return_message +=
+                cstring_to_asciihex_string("this is a "
                                            "non-interactive debug session, "
                                            "cannot get permission to debug "
                                            "processes.");
```
this is ended with a '.', we pick how we want to do it and do it consistently

jasonmolenda commented:
`CreateEXXPacket("E96", "no such process");` that's a good idea, I'll rewrite to use a helper method like that.

jasonmolenda committed:
Use new RNBRemote::SendErrorPacket method in debugserver

Have all error packet replies go through `SendErrorPacket()` with
an optional string explaining the error, and send the optional
string if extended error message reporting is enabled.

Also found that RNBRemote::HandlePacket_D was not sending properly
formatted error packets if there was an error; fixed that.

jasonmolenda committed:
Keep the RNBRemote::PacketEnum table hand-laid-out

jasonmolenda commented:
Going back to this PR, @bulbazord @clayborg and I agreed that having a single method for sending error reply packets, which correctly encode the error string it if is available and extended-error-responses has been enabled, would be worth modifying all the error returns, so I did that.  I found an additional bad error response in HandlePacket_D where it did not include hex digits on its error response.

I think we're good to land this now.

bulbazord reviewed (approved):
Looks great!!! 😄 Thanks for fixing the bug and gardening!

jasonmolenda committed:
Merge branch 'main' into fix-error-string-returns-from-debugserver
